README.md
# react-tenor

[![Build Status](https://github.com/CultureHQ/react-tenor/workflows/Main/badge.svg)](https://github.com/CultureHQ/react-tenor/actions)
[![Package Version](https://img.shields.io/npm/v/react-tenor.svg)](https://www.npmjs.com/package/react-tenor)

A React component for selected GIFs from [Tenor](https://tenor.com/gifapi).

## Getting started

First, add `react-tenor` to your `package.json` `dependencies`, then install using either `npm install` or `yarn install`. Then, get your API key from tenor. Finally, you can add the selector component by adding:

```jsx
<Tenor token="your-token-here" onSelect={result => console.log(result)} />
```

### Styles

To get the styles, be sure it import `react-tenor/dist/styles.css` into your application. You can style it appropriately for your app by overriding the CSS classes used internally. They are listed in [`styles.css`](src/styles.css).

### Props

Below is a list of all of the props you can pass to the `Tenor` component.

| Name                | Type             | Default              | Description                                                                                                                                  |
| ------------------- | ---------------- | -------------------- | -------------------------------------------------------------------------------------------------------------------------------------------- |
| `autoFocus`         | `boolean`        | `false`              | Indicates that the search bar of the component should request focus when it first mounts.                                                    |
| `base`              | `string`         | `"api.tenor.com/v1"` | The base of the API that this component hits.                                                                                                |
| `contentFilter`     | `string`         | `"mild"`             | The content filter that gets passed up to tenor. See the [tenor API docs](https://tenor.com/gifapi/documentation#contentfilter) for details. |
| `contentRef`        | `Ref`            | `null`               | A ref to the `div` that the `Tenor` component renders.                                                                                       |
| `defaultResults`    | `boolean`        | `false`              | Indicates that the component should automatically search for trending results if the search input is empty.                                  |
| `initialSearch`     | `string`         | `""`                 | The starting value of the search bar.                                                                                                        |
| `limit`             | `number`         | `12`                 | The number of results to return for each search.                                                                                             |
| `locale`            | `string`         | `"en_US"`            | The locale that gets passed up to tenor. See the [tenor API docs](https://tenor.com/gifapi/documentation) for details.                       |
| `mediaFilter`       | `string`         | `"minimal"`          | The media filter that gets passed up to tenor. See the [tenor API docs](https://tenor.com/gifapi/documentation) for details.                 |
| `onSelect`          | `Result => void` |                      | A callback for when the user selects a GIF.                                                                                                  |
| `searchPlaceholder` | `string`         | `"Search Tenor"`     | The placeholder that is applied to the search input field.                                                                                   |
| `token`             | `string`         |                      | The tenor API token. See the [tenor API docs](https://tenor.com/gifapi/documentation) for details.                                           |

## Testing locally

You can run the tests by running `yarn test` and lint by running `yarn lint`. You can run the local server by running `yarn start` which will start the docs server on `http://localhost:8080`.

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/CultureHQ/react-tenor.

## License

The code is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.2.0] - 2020-03-13

### Added

- The `limit` prop to support setting how many results to return.

## [2.1.1] - 2019-09-23

### Changed

- Correctly pass down the `searchPlaceholder` prop.

## [2.1.0] - 2019-09-19

### Added

- The `searchPlaceholder` prop for configuring the placeholder on the search input.

## [2.0.0] - 2019-09-13

### Changed

- Renamed the `safesearch` param on the component to `contentFilter`.

## [1.5.0] - 2019-09-12

### Added

- The `locale`, `mediaFilter`, and `safesearch` props.

### Changed

- Switch to TypeScript for development.
- Do not fire off autocomplete and suggestion requests when the search is blank but you have `defaultResults` selected.

## [1.4.0] - 2019-08-09

### Added

- The `autoFocus` prop to focus on the input when the component mounts.
- The `defaultResults` prop to display trending results in empty search.

### Changed

- No longer rely on `fetch` being available.
- [INTERNAL] Refactor Client.js to be simpler.
- Ensure `pointer` cursor on pagination controls.
- Align typeahead text with current text.

## [1.3.2] - 2019-06-07

### Changed

- Use the `prepublishOnly` npm script so that the `dist` directory does not need to be checked into the repository.

## [1.3.1] - 2019-06-07

### Changed

- Properly rebuild dist.

## [1.3.0] - 2019-06-06

### Added

- The `initialSearch` prop.

## [1.2.1] - 2019-06-05

### Changed

- Rebuilt dist with correct capitalization of file names.

## [1.2.0] - 2019-05-22

### Added

- Switched to using `@culturehq/scripts` for development.
- Added `:focus` styles for relevant components for better keyboard support.

## [1.1.0] - 2018-10-03

### Added

- Type ahead through the use of Tenor's autocomplete feature.
- Suggestions are now rendered through the user of Tenor's suggestions feature.
- Pagination is now supported. Additionally you can use the meta-key plus left arrow and right arrow to go between pages.

### Changed

- Now encoding the URI being sent to tenor to ensure it's a valid URL.
- If you click outside the component, the component now knows about that click and closes the selector.
- The search bar is now of type "search" which builds in some nicities from the browsers that support it.
- Added better accessibility support by properly naming the GIF buttons.
- Load preview GIFs in the background and then set them to fade in once they are loaded.

## [1.0.0] - 2018-09-26

### Added

- The optional `contentRef` prop to get access to the actual div that is being rendered.
- The optional `base` prop that will specify the base of the API for the search URLs that are generated.
- The `focus()` function on the main component to allow consumers to focus into the input field.

### Changed

- Removed the style import by default. It's now up to the consumers of this package to import `react-tenor/dist/styles.css` into their applications. This avoids a lot of weird webpack bugs.

## [0.2.0] - 2018-07-18

### Changed

- Don't build the final distribution with `webpack`, just use `babel`.
- Rename `example` to `docs` so we can publish to github pages.

[unreleased]: https://github.com/CultureHQ/react-tenor/compare/v2.2.0...HEAD
[2.2.0]: https://github.com/CultureHQ/react-tenor/compare/v2.1.1...v2.2.0
[2.1.1]: https://github.com/CultureHQ/react-tenor/compare/v2.1.0...v2.1.1
[2.1.0]: https://github.com/CultureHQ/react-tenor/compare/v2.0.0...v2.1.0
[2.0.0]: https://github.com/CultureHQ/react-tenor/compare/v1.5.0...v2.0.0
[1.5.0]: https://github.com/CultureHQ/react-tenor/compare/v1.4.0...v1.5.0
[1.4.0]: https://github.com/CultureHQ/react-tenor/compare/v1.3.2...v1.4.0
[1.3.2]: https://github.com/CultureHQ/react-tenor/compare/v1.3.1...v1.3.2
[1.3.1]: https://github.com/CultureHQ/react-tenor/compare/v1.3.0...v1.3.1
[1.3.0]: https://github.com/CultureHQ/react-tenor/compare/v1.2.1...v1.3.0
[1.2.1]: https://github.com/CultureHQ/react-tenor/compare/v1.2.0...v1.2.1
[1.2.0]: https://github.com/CultureHQ/react-tenor/compare/v1.1.0...v1.2.0
[1.1.0]: https://github.com/CultureHQ/react-tenor/compare/v1.0.0...v1.1.0
[1.0.0]: https://github.com/CultureHQ/react-tenor/compare/v0.2.0...v1.0.0
[0.2.0]: https://github.com/CultureHQ/react-tenor/compare/2d68b4...v0.2.0


src/Search.js
import React from "react";

import Result from "./Result";

const AutoComplete = ({ autoComplete, search }) => {
  const prefix = search.toLowerCase().replace(/\s/g, "");
  const typeahead = autoComplete.toLowerCase().replace(prefix, "");

  return (
    <div className="react-tenor--autocomplete">
      <span aria-hidden>{search}</span>
      {typeahead}
    </div>
  );
};

const Spinner = () => (
  <svg className="react-tenor--spinner" viewBox="0 0 1024 1024">
    <path d="M959.6 452.2c-2.8-17.4-6.2-34.6-10.6-51.6-5.6-21.6-12.8-43-21.6-63.6-17.8-42.4-42.2-82-71.8-117.2-32-37.8-70.6-70.4-113.4-95.4-42.2-24.8-88.2-42.4-136.2-52.2-24.8-5-49.8-8-75.2-8.2-19.8-0.2-39.6 0.6-59.2 2.4-51 5-101.4 19.2-147.8 41-39.8 18.8-76.8 43.2-109.6 72.4s-61.4 63.2-84.4 100.6c-25.4 41.6-44.4 87-54.8 134.6-8.4 38-12.4 77.2-10.4 116.2 1.8 37.8 7.6 75.6 19 111.8 7.2 23 15.8 45.4 26.6 67.2 10.6 21.4 23 42 36.8 61.4 27.6 38.6 61.2 72.8 99.6 101 39.2 29 83.4 51.4 129.8 66.2 48.4 15.4 99.8 22.6 150.6 20.8 49.6-1.6 98.8-11.2 145.2-29 44.6-17.2 86.4-41.8 123-72.6 18.4-15.6 34.8-33.2 50.2-51.8 15.6-18.8 29.6-38.6 41.2-60 10-18.4 18.4-37.6 25.6-57 3.6-9.6 7-19.2 9.8-29.2 3-10.6 5.2-21.6 7.2-32.4 3-17 4.2-34.6 2.6-51.8-1.4 7.6-2.6 15-4.4 22.4-2.2 8.6-5 17-8.2 25.2-6.4 17.4-14.4 34.2-22 51-9.8 21.4-21.2 41.8-33.6 61.6-6.4 10.2-13 20.2-20.2 29.8s-15.4 18.8-23.6 27.8-34.2 34.4-54 48.8c-20.2 14.8-41.6 27.8-64.2 38.6-45.2 22-94.6 35.2-144.6 39.6-51.2 4.4-103.4-0.6-152.6-15.2-46.8-13.8-91.2-36.2-130.2-65.6-37.8-28.6-70.6-63.8-96.4-103.6-27-40.6-45.6-86.4-55.8-134.2-2.6-12.4-4.6-25.2-6-37.8-1.2-10.8-2.2-21.8-2.6-32.8-0.6-22.6 0.8-46 4.2-68.4 7.4-49.2 23.4-96.6 48.2-139.8 22-38.6 50.6-73.4 84.2-102.8 33.6-29.6 72-53.4 113.6-70.2 24-9.8 49.2-17 74.8-21.8 13-2.4 26-4.4 39-5.4 6.4-0.6 12.6-0.6 19-1.2 2.6 0 5.2 0.2 7.8 0.2 43.4-0.8 87 4.8 128.4 17.8 44.6 14 86.6 36.6 123 66 38.2 30.8 70 68.8 94 111.6 20.4 36.4 35 75.6 43.8 116.4 2 9.4 3.6 18.8 5.2 28.2 1.4 8.6 6.2 16.6 13.6 21.4 15.6 10.4 37.4 3.4 45-13.4 2.6-5.8 3.4-12 2.4-17.8z" />
  </svg>
);

const SearchBar = ({
  autoComplete, inputRef, search, searching, onSearchChange, onSearchKeyDown
}) => (
  <div className="react-tenor--search-bar">
    <input
      ref={inputRef}
      aria-label="Search"
      className="react-tenor--search"
      type="search"
      value={search}
      onChange={onSearchChange}
      onKeyDown={onSearchKeyDown}
      placeholder="Search Tenor"
    />
    {autoComplete && search && (
      <AutoComplete autoComplete={autoComplete} search={search} />
    )}
    {searching && <Spinner />}
  </div>
);

const Suggestion = ({ suggestion, onSuggestionClick }) => {
  const onClick = () => onSuggestionClick(suggestion);

  return <button type="button" onClick={onClick}>{suggestion}</button>;
};

const Suggestions = ({ suggestions, onSuggestionClick }) => (
  <div className="react-tenor--suggestions">
    {suggestions.map(suggestion => (
      <Suggestion
        key={suggestion}
        suggestion={suggestion}
        onSuggestionClick={onSuggestionClick}
      />
    ))}
  </div>
);

const Results = ({ results, onPageLeft, onPageRight, onSelect }) => (
  <div className="react-tenor--results">
    {results.map(result => (
      <Result key={result.id} result={result} onSelect={onSelect} />
    ))}
    <PageControl direction="left" onClick={onPageLeft} />
    <PageControl direction="right" onClick={onPageRight} />
  </div>
);

const PageControl = ({ direction, onClick }) => (
  <button
    aria-label={`Page ${direction}`}
    className={`react-tenor--page-${direction}`}
    type="button"
    onClick={onClick}
  >
    <div />
  </button>
);

const Search = ({
  autoComplete, contentRef, inputRef, onPageLeft, onPageRight, onSearchChange,
  onSearchKeyDown, onSuggestionClick, onSelect, results, search, searching,
  suggestions
}) => {
  let classList = "react-tenor";
  if (suggestions.length > 0 || results.length > 0) {
    classList = `${classList} react-tenor-active`;
  }

  return (
    <div className={classList} ref={contentRef}>
      <SearchBar
        autoComplete={autoComplete}
        inputRef={inputRef}
        search={search}
        searching={searching}
        onSearchChange={onSearchChange}
        onSearchKeyDown={onSearchKeyDown}
      />
      {suggestions.length > 0 && (
        <Suggestions
          suggestions={suggestions}
          onSuggestionClick={onSuggestionClick}
        />
      )}
      {results.length > 0 && (
        <Results
          results={results}
          onPageLeft={onPageLeft}
          onPageRight={onPageRight}
          onSelect={onSelect}
        />
      )}
    </div>
  );
};

export default Search;
